File: android/src/main/java/com/woot/notification/extensions/SQLiteHandler.kt
package com.woot.notification.extensions

import android.content.Context
import android.content.pm.PackageManager
import com.getcapacitor.JSArray
import com.woot.notification.extensions.sqliteHelper.SQLiteDatabaseHelper
import java.lang.Exception


class SQLiteHandler(private var context: Context) {
    private val encrypted = false
    private val mode = "no-encryption"
    private val dbVersion = 1
    private val tableName = "notification_extensions_filter"
    private lateinit var mdb: SQLiteDatabaseHelper

    fun openDB(): Boolean {
        context.packageManager
                .getApplicationInfo(context.packageName, PackageManager.GET_META_DATA)
                .apply {
                    val dbName = metaData.getString("com.woot.notification.extensions.local_database_name")
                    mdb = SQLiteDatabaseHelper(
                            context,
                            dbName!! + "SQLite.db",
                            encrypted,
                            mode,
                            dbVersion = dbVersion
                    )
                    return mdb.isOpen
                }
    }

    @Throws(Exception::class)
    fun createFilterTable() {
        if (::mdb.isInitialized) {
            val statement = StringBuilder("CREATE TABLE IF NOT EXISTS ")
                    .append(tableName)
                    .append(" (id INTEGER PRIMARY KEY NOT NULL, key TEXT NOT NULL UNIQUE, value TEXT);")
                    .toString()
            val createTableSQL: Array<String> = arrayOf(statement)
            mdb.execSQL(createTableSQL)
        } else {
            throw Exception("Local database not opened yet.")
        }
    }

    @Throws(Exception::class)
    fun getAllFilters(): JSArray {
        if (::mdb.isInitialized) {
            val statement = StringBuilder("SELECT * FROM")
                    .append(tableName)
                    .toString()
            return mdb.querySQL(statement, null)
        } else {
            throw Exception("Local database not opened yet.")
        }
    }

    @Throws(Exception::class)
    fun getTimeFilter(): JSArray {
        if (::mdb.isInitialized) {
        val statement = StringBuilder("SELECT * FROM ")
                .append(tableName)
                .append(" WHERE key IN ('filter_start_from', 'filter_end_at', 'is_time_filter_on');")
                .toString()
        return mdb.querySQL(statement, null)
        } else {
            throw Exception("Local database not opened yet.")
        }
    }

    @Throws(Exception::class)
    fun getFilters(): JSArray {
        if (::mdb.isInitialized) {
            val statement = StringBuilder("SELECT * FROM ")
                    .append(tableName)
                    .append(" WHERE key NOT IN ('filter_start_from', 'filter_end_at', 'is_time_filter_on');")
                    .toString()
            return mdb.querySQL(statement, null)
        } else {
            throw Exception("Local database not opened yet.")
        }
    }

    @Throws(Exception::class)
    fun insertTimeFilter(startFrom: String, endAt: String): Map<String, Any> {
        if (startFrom.split(':').size != 2 || endAt.split(':').size != 2) {
            return mapOf("success" to false, "reason" to "Invalid time format")
        }
        if (::mdb.isInitialized) {
            val values: Array<String> = arrayOf(
                    "('filter_start_from', '$startFrom')",
                    "('filter_end_at', '$endAt')",
                    "('is_time_filter_on', 'true')"
            )
            val statement = StringBuilder("INSERT OR REPLACE INTO ")
                    .append(tableName)
                    .append(" (key, value) VALUES ")
                    .append(values.joinToString(", "))
                    .append(';')
                    .toString()
            mdb.execSQL(arrayOf(statement))
            return mapOf("success" to true)
        } else {
            throw Exception("Local database not opened yet.")
        }
    }

    @Throws(Exception::class)
    fun removeTimeFilter(): Map<String, Any> {
        if (::mdb.isInitialized) {
            val statement = StringBuilder("INSERT OR REPLACE INTO ")
                    .append(tableName)
                    .append(" (key, value) VALUES ")
                    .append("('is_time_filter_on', 'false');")
                    .toString()
            mdb.execSQL(arrayOf(statement))
            return mapOf("success" to true)
        } else {
            throw Exception("Local database not opened yet.")
        }
    }

    @Throws(Exception::class)
    fun insertFilter(key: String): Map<String, Any> {
        return if (::mdb.isInitialized) {
            try {
                val statement = StringBuilder("INSERT OR REPLACE INTO ")
                        .append(tableName)
                        .append(" (key, value) VALUES ")
                        .append("('$key', 'false');")
                        .toString()
                mdb.execSQL(arrayOf(statement))
                mapOf("success" to true)
            } catch (exception: Throwable) {
                mapOf("success" to false, "reason" to "Unexpected error occurred while insert filter '$key'")
            }
        } else {
            throw Exception("Local database not opened yet.")
        }
    }

    @Throws(Exception::class)
    fun removeFilter(key: String): Map<String, Any> {
        return if (::mdb.isInitialized) {
            try {
                val statement = StringBuilder("INSERT OR REPLACE INTO")
                        .append(tableName)
                        .append(" (key, value) VALUES ")
                        .append("('$key', 'true');")
                        .toString()
                mdb.execSQL(arrayOf(statement))
                mapOf("success" to true)
            } catch (exception: Throwable) {
                mapOf("success" to false, "reason" to "Unexpected error occurred while insert filter '$key'")
            }
        } else {
            throw Exception("Local database not opened yet.")
        }
    }
}

File: ios/Plugin/SQLiteHandler.swift
//
//  File.swift
//  NotificationExtensions
//
//  Created by mjLee on 2020/12/29.
//

import Foundation
import SQLite3

class SQLiteHandler {
    var db: OpaquePointer?
    var isDbOpened = false
    private let tableName = "notification_extensions_filter"
    init() {
        openDB()
    }
    
    // MARK: - Open Database
    func openDB() {
        let fileURL = try! FileManager.default.url(for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: false).appendingPathComponent("TestDatabase.sqlite")
            
        if sqlite3_open(fileURL.path, &db) != SQLITE_OK {
            print("open DB Failed")
            isDbOpened = false
        } else {
            print("open DB Succeed")
            isDbOpened = true
        }
    }
    
    // MARK: - Create Filter Table
    func createFilterTable() -> Void {
        if (!isDbOpened) {
            print("Local database not opened yet.")
            return
        }
        var sql = "CREATE TABLE IF NOT EXISTS "
            sql += tableName
            sql += "(id INTEGER PRIMARY KEY NOT NULL, key TEXT NOT NULL UNIQUE, value TEXT);"
        if sqlite3_exec(db, sql, nil, nil, nil) != SQLITE_OK {
            let errmsg = String(cString: sqlite3_errmsg(db)!)
            print("error creating table: \(errmsg)")
        }
    }
    
    // MARK: - Get Time Filters
    func getTimeFilter() -> Array<[String: Any]> {
        if (!isDbOpened) {
            print("Local database not opened yet.")
            return []
        }
        
        var sql = "SELECT * FROM "
            sql += tableName
            sql += " WHERE key IN ('filter_start_from', 'filter_end_at', 'is_time_filter_on');"

        return runSelectQuery(sql: sql)
    }
    
    // MARK: - Get Non-Time Filters
    func getFilters() -> Array<[String: Any]> {
        if (!isDbOpened) {
            print("Local database not opened yet.")
            return []
        }
        var sql = "SELECT * FROM "
            sql += tableName
            sql += " WHERE key NOT IN ('filter_start_from', 'filter_end_at', 'is_time_filter_on');"
        
        return runSelectQuery(sql: sql)
    }
    
    // MARK: - Add Time Filter
    func insertTimeFilter(startFrom: String, endAt: String) -> [String: Any] {
        if (!isDbOpened) {
            return ["success": false, "reason": "Local database not opened yet."]
        }
        
        if (startFrom.components(separatedBy: ":").count != 2 || endAt.components(separatedBy: ":").count != 2) {
            return ["success": false, "reason": "Invalid time format"]
        }

        var sql = "INSERT OR REPLACE INTO "
            sql += tableName
            sql += " (key, value) VALUES "
//            sql += "('filter_start_from', '09:10'), "
//            sql += "('filter_end_at', '23:20'), "
            sql += "('filter_start_from', '\(startFrom)'), "
            sql += "('filter_end_at', '\(endAt)'), "
            sql += "('is_time_filter_on', 'true');"

        return runDataChangingQuery(sql: sql)
    }
    
    // MARK: - Remove Time Filter
    func removeTimeFilter() -> [String: Any]  {
        if (!isDbOpened) {
            return ["success": false, "reason": "Local database not opened yet."]
        }

        var sql = "INSERT OR REPLACE INTO "
            sql += tableName
            sql += " (key, value) VALUES "
            sql += "('is_time_filter_on', 'false');"
        
        return runDataChangingQuery(sql: sql)
    }
    
    // MARK: - Add Non-Time Filter
    func insertFilter(key: String) -> [String: Any] {
        if (!isDbOpened) {
            return ["success": false, "reason": "Local database not opened yet."]
        }
        
        var sql = "INSERT OR REPLACE INTO "
            sql += tableName
            sql += " (key, value) VALUES "
            sql += "('\(key)', 'false');"
        
        return runDataChangingQuery(sql: sql)
    }
    
    // MARK: - Remove Non-Time Filter
    func removeFilter(key: String) -> [String: Any] {
        if (!isDbOpened) {
            return ["success": false, "reason": "Local database not opened yet."]
        }
        
        var sql = "INSERT OR REPLACE INTO "
            sql += tableName
            sql += " (key, value) VALUES "
            sql += "('\(key)', 'true');"
        
        return runDataChangingQuery(sql: sql)
    }
    
    // run insert, update, delete query
    func runDataChangingQuery(sql: String) -> [String: Any] {
        var statement: OpaquePointer? = nil;
        var errmsg: String
        if sqlite3_prepare(db, sql, -1, &statement, nil) == SQLITE_OK {
            if sqlite3_step(statement) == SQLITE_DONE {
                return ["success": true]
            } else {
                errmsg = String(cString: sqlite3_errmsg(db)!)
            }
        } else {
            errmsg = String(cString: sqlite3_errmsg(db)!)
            print("get timeFilter: \(errmsg)")
        }
        
        return ["success": false, "reason": errmsg]
    }
    
    // run select query
    func runSelectQuery(sql: String) -> Array<[String: Any]> {
        var statement: OpaquePointer? = nil;
        if sqlite3_prepare(db, sql, -1, &statement, nil) == SQLITE_OK {
            var resultArray = Array<[String: Any]>()
            while sqlite3_step(statement) == SQLITE_ROW {
                let id = sqlite3_column_int(statement, 0) as Int32
                let key: String = String(cString: sqlite3_column_text(statement, 1))
                let value: String = String(cString: sqlite3_column_text(statement, 2))
                resultArray.append(["id": id, "key": key, "value": value])
            }
            return resultArray
        } else {
            let errmsg = String(cString: sqlite3_errmsg(db)!)
            print("get timeFilter: \(errmsg)")
            return []
        }
    }
}
